### PR TITLE
Replace macro with function for `array_position` and `array_positions`

### LINKED ISSUE #8145: Generalized `array_position` to get at most n element and start from certain position.
### Is your feature request related to a problem or challenge?
While cleanup `array_position` I find we can extend it have better feature easily #7988 

# Background
`array_position` is the early function that follows on Postgresql.

There are three types of interface now.
```
array_position(array, element) 
array_position(array, element, index)
array_positions(array, element)
```

`array_position(array, element) ` is the basic one. 
> For example `array_position([1,2,3], 3)` return 3 (1-indexed).

`array_position(array, element, index)`  finds the element `start_from` index. 
> For example `array_position([1,2,3,4,3], 3, 4) ` return 5.

`array_positions(array, element)` return all the occurrence of the elements. 
> For example `array_positions([1,2,2,3,3,3], 2)` return [2,3].

# Problem
We are not able to find n element start from certain index.

# Proposal 1 Extend the current interface
`array_position(array, element, Option<index>, Option<n>)` that able us to find n elements start from `index` and
`array_positions(array, element, Option<index>)` that able us to find all the elements start from `index`.

# Proposal 2 Introduce `array_position_n`
`array_position(array, element, Option<index>)` that able us to find 1 element start from `index`.
`array_position_n(array, element, n, Option<index>)` that able us to find n elements start from `index`.
`array_positions(array, element, Option<index>)` that able us to find all the elements start from `index`.

I think either of two are good for me, so I would like to collect feedbacks from yours.

# Others than PostgreSQL
## [Duckdb](https://duckdb.org/docs/sql/functions/nested.html)
`list_position(list, element)` has only the basic one.

## Clickhouse
None

## [Spark](https://sparkbyexamples.com/spark/spark-sql-array-functions/)
`array_position(column: Column, value: Any) ` has only the basic one. 

## [Azure](https://learn.microsoft.com/en-us/azure/databricks/sql/language-manual/functions/array_position)
`array_position(array, element)` has only the basic one.

# Note
I had not yet find other SQL have the similar function, so I'm not sure if this feature is helpful or not.

### Describe the solution you'd like

Extend `array_position`

### Describe alternatives you've considered

Keep it as it is.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Part of #7988
Closes #8145

### CODE REVIEW

**Review by alamb — APPROVED** [review]
This certainly seems like an improvement to me. I think we can probably make it even more efficient but doing that as a follow on PR would make a lot of sense too. Thank you @jayzhan211 

cc @Veeupup  in case you have some other thoughts

File: datafusion/physical-expr/src/array_expressions.rs [hunk 1]
```diff
@@ -953,114 +1025,68 @@ fn general_list_repeat(
...
-            .map(|(arr, el)| match arr {
-                Some(arr) => {
-                    let child_array = downcast_arg!(arr, $ARRAY_TYPE);
-                    let res = child_array
-                        .iter()
-                        .enumerate()
-                        .filter(|(_, x)| *x == el)
-                        .flat_map(|(i, _)| Some((i + 1) as u64))
-                        .collect::<UInt64Array>();
+fn general_position<OffsetSize: OffsetSizeTrait>(
```
Comment (Veeupup) [inline datafusion/physical-expr/src/array_expressions.rs:1060]: if we only use i32 as OffsetSize, do we really need the Generics?

**Review by Veeupup — APPROVED** [review]
LGTM : )

File: datafusion/physical-expr/src/array_expressions.rs (hunk 1 above)
Comment (jayzhan211) [reply datafusion/physical-expr/src/array_expressions.rs:1060]: no, but we will need it if we want to extend it for large list, I just lazy to remove it for now :)

Comment (Weijun-H) [reply datafusion/physical-expr/src/array_expressions.rs:1060]: Nice extension 👍

**Review by Weijun-H — APPROVED** [review]

**Pull request merged**